test/asami/query_internals_test.cljc

```diff
@@ -363,4 +363,61 @@
       (is (vartest? p))
       (is (vartest? v)))))
 
+(deftest query-validator-test
+  (testing "valid queries"
+    ;; This example represents the uncommon, but legal, scenario
+    ;; wherein some variables appearing in the :find clause are only
+    ;; present in the :in clause. Consider the following.
+    ;;
+    ;;     (defn my-query [param]
+    ;;       (asami.core/q '{:find [,,, ?param ,,,]
+    ;;                       :in [$ ?param] ,,,}
+    ;;                     db
+    ;;                     param))
+    ;;
+    ;; In this case, a user may rely on substitution rather than
+    ;; inserting the value of param in a post processing step, etc.
+    (is (= '{:find [?x ?y], :in [$ ?x], :where [[?y ?z]]}
+           (q/query-validator '{:find [?x ?y] :in [$ ?x] :where [[?y ?z]]}))))
+
+  (testing "invalid queries"
+    (is (thrown-with-msg? ExceptionInfo #"Missing ':find' clause"
+                          (q/query-validator {})))
+
+    (is (thrown-with-msg? ExceptionInfo #"Missing ':find' clause"
+                          (q/query-validator {:find []})))
```

Comment:
Is an empty `:find` illegal? I can see it being useful, as in:

```clojure
(when (seq (q '[:find :where [?e :type :data]] db))
  (println "Data exists"))
```

Or should this be a query that starts with `:exists` instead of `:find`? Just thinking out loud here...